### PR TITLE
Add TransactionInterceptorFactoryBean

### PR DESCRIPTION
**Pull request by garyrussell**
http://stackoverflow.com/questions/39304371/how-properly-define-spring-integration-jpaoutboundgateway-in-java-configuration/39308078?noredirect=1#comment65967496_39308078

### CODE REVIEW

File: src/main/java/org/springframework/integration/dsl/transaction/TransactionInterceptorFactoryBean.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+			transactionAttribute.setTimeout(this.timeout);
+			transactionAttribute.setReadOnly(this.readOnly);
+			attSource.setTransactionAttribute(transactionAttribute);
+		}
+		else {
+			attSource.setTransactionAttribute(this.transactionAttribute);
+		}
+		if (this.transactionManager == null) {
+			this.transactionManager = this.getBeanFactory().getBean("transactionManager",
+					PlatformTransactionManager.class);
```
Comment (artembilan) [inline src/main/java/org/springframework/integration/dsl/transaction/TransactionInterceptorFactoryBean.java]: I'm intended to disagree. See `TransactionAspectSupport.determineTransactionManager()`:
```
else if (StringUtils.hasText(this.transactionManagerBeanName)) {
	return determineQualifiedTransactionManager(this.transactionManagerBeanName);
}
else {
	PlatformTransactionManager defaultTransactionManager = getTransactionManager();
	if (defaultTransactionManager == null) {
		defaultTransactionManager = this.beanFactory.getBean(PlatformTransactionManager.class);
		this.transactionManagerCache.putIfAbsent(
		DEFAULT_TRANSACTION_MANAGER_KEY, defaultTransactionManager);
	}
	return defaultTransactionManager;
}
```
I don't say that we should allow to provide `transactionManagerBeanName`, but at the same time we shouldn't be so strict for the `defaultTransactionManager`. 
In other words let's rely on the `TransactionAspectSupport` logic and don't do any `getBean()` here!

Therefore this class can be freed from the `AbstractFactoryBean` oppression, and may just turn to the beautiful butterfly - raw Builder class :smile: .

Good work anyway!

**Comment by artembilan** [comment]
I wonder if we can replace and enhance an existing logic in the:
```
public PollerSpec transactional(PlatformTransactionManager transactionManager) {
	return this.advice(new TransactionInterceptor(transactionManager, new MatchAlwaysTransactionAttributeSource()));
}
```
I think with the simple Builder that can look pretty.

**Comment by garyrussell** [comment]
> I think with the simple Builder that can look pretty.

We can do that when we merge the DSL into core.

**Comment by garyrussell** [comment]
Forget that - I thought you were talking about `@Poller`

File: src/main/java/org/springframework/integration/dsl/core/PollerSpec.java [hunk 2]
```diff
@@ -109,9 +110,21 @@ public PollerSpec advice(Advice... advice) {
 	 * provided {@code PlatformTransactionManager} for the {@code pollingTask}.
 	 * @param transactionManager the {@link PlatformTransactionManager} to use.
 	 * @return the spec.
+	 * @deprecated - since 1.2 use {@link #transactional(TransactionInterceptor)}
 	 */
+	@Deprecated
```
Comment (artembilan) [inline src/main/java/org/springframework/integration/dsl/core/PollerSpec.java]: I don't think so. 
There is no problems with this method.
It is a good choice to be based on just a provided `PlatformTransactionManager`.
I also see one more choice as a argumentless `.transactional()`, where I don't see reason to require `new TransactionInterceptorBuilder().build()` from end-user where he/she would like to rely on the default for everything.
WDYT?

File: src/main/java/org/springframework/integration/dsl/transaction/TransactionInterceptorBuilder.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+ */
+public class TransactionInterceptorBuilder {
+
+	private Propagation propagation = Propagation.REQUIRED;
+
+	private Isolation isolation = Isolation.DEFAULT;
+
+	private int timeout = -1;
+
+	private boolean readOnly = false;
```
Comment (artembilan) [inline src/main/java/org/springframework/integration/dsl/transaction/TransactionInterceptorBuilder.java]: I'd say we don't need these properties here and just have `DefaultTransactionAttribute final` instance and delegate from the setters.
OTOH, I see that we even can have `final TransactionInterceptor` and use its getters to modify its internals.
WDYT?

File: src/main/java/org/springframework/integration/dsl/core/PollerSpec.java (hunk 2 above)
Comment (garyrussell) [inline src/main/java/org/springframework/integration/dsl/core/PollerSpec.java]: OK; makes sense; but I guess we need to explain in the javadocs it just gets default props.

**Comment by garyrussell** [comment]
For some reason I can't add a line note to your last comment - ok with me.

**Comment by garyrussell** [comment]
For some reason I can't add a line note to your last comment - ok with me.

**Comment by artembilan** [comment]
Polishing for your confirmation: https://github.com/artembilan/spring-integration-java-dsl/commit/bb6ffc1ab1bf5728f29713616fdca1f532de7a5b

**Comment by garyrussell** [comment]
Polish is good and travis agrees.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration-java-dsl/commit/846cd90d6d1d57d4d87e2b47328945a35620ce1a

**Pull request closed without merging**